decisions/websre/0004-use-1-kubernetes-cluster-per-environment.md

```diff
@@ -0,0 +1,58 @@
+* Development clusters can optionally include automatic infrastructure or platform updates, as a first place to review these changes and their effects, with a goal of promoting these updates through stage and prod if feasible;
+* Only use reviewed & approved, stable infrastructure releases on the Stage cluster, and deploy those infrastructure changes when able to relatively quickly move the changes from stage into prod to avoid divergent results from staging and production applications;
+* Only use reviewed & approved, production-ready stable application & infrastructure releases on the Production cluster;
+* Understand that the purpose of the Stage cluster is to validate infrastructure & application code headed to Production. As such, we expect that deployment on the Stage cluster is a step on the release & deployment pipelines for an application or infrastructure moving to Production.
+
+This ADR does not include the following:
+* How the applications within each cluster name their namespaces & other application-specific Kubernetes resources (e.g. keeping "discourse_prod" namespace in a production cluster is redundent, but also very explicit, and is left up to the SREs working on the service);
+* An immediate push to migrate all existing services to this pattern, as it will cause a lot of churn. Instead, this should be used going forward when doing migrations or starting new services;
+* Explicit declarations on SLOs or SLAs for these environments, beyond the decisions that Development cluster(s) does not have round-the-clock support;
+* Any expectation on multi-tenant clusters versus single-tenant clusters - the pattern for environments should apply in either case;
+* An expectation that a development cluster exists - this is left up to a service's need(s) and SREs' determination;
+* Decisions on approaches to non-peristent clusters, e.g. local development or CI pipeline environments;
+* Decisions on how to implement 1 cluster per environment - this is left to the SREs' implementation decisions.
+
+## Consequences
+
+Pros:
+* We have a pattern that supports a meaningful & separate staging environment, especially for infrastructure changes where testing their impact on applications is trickier in cases where dev, staging, and prod applications all live in the production cluster;
+* There is a clearer boundary between environments within the infrastructure;
+
+Cons:
+* There will need to be work to bring our current services into alignment, including a fair amount of migrations and infrastructure updates, hence the decision above that this ADR is considered a pattern to involve when creating a new service or performing a migration;
+* Admin or Developer+ users of the Kubernetes clusters will have to hop between clusters to compare environment states, logs, etc. This is mitigated somewhat by Kubernetes command line tooling, though.
+
```

Comment:
Another downside is a higher overhead with more clusters (more control planes, less efficient bin packing on nodes).